ecl/hqlcpp/hqlhtcpp.cpp

```diff
@@ -8864,6 +8881,97 @@ ABoundActivity * HqlCppTranslator::doBuildActivityRemote(BuildCtx & ctx, IHqlExp
 }
 
 
+//---------------------------------------------------------------------------------------------------------------------
+
+ABoundActivity * HqlCppTranslator::doBuildActivityEmbed(BuildCtx & ctx, IHqlExpression * expr, bool isRoot)
+{
+    CIArray bound;
+    IHqlExpression * funcdef = queryFuncdef(expr);
+    IHqlExpression * formals = funcdef->queryChild(1);
+    ForEachChild(iInput, formals)
+    {
+        IHqlExpression * input = formals->queryChild(iInput);
+        if (!isStreamingDataset(input))
+            break;
+
+        IHqlExpression * actual = expr->queryChild(iInput);
+        assertex(actual);
+        bound.append(*buildCachedActivity(ctx, actual));
+    }
+
+    ThorActivityKind kind;
+    if (expr->isDataset())
+        kind = bound.empty() ? TAKexternalsource : TAKexternalprocess;
+    else
+        kind = TAKexternalsink;
+
+    Owned<ActivityInstance> instance = new ActivityInstance(*this, ctx, kind, expr, "External");
+
+    //Substitute the parameters into the external call/embed definition, so that any attributes that depend on the arguments
+    //are expanded.
+    OwnedHqlExpr expandedCall = expandOutOfLineFunctionCall(expr);
+
+    //The setting for the local attribute allows the localness to be configured for the activity
+    IHqlExpression * localAttr = expandedCall->queryAttribute(localAtom);
+    if (localAttr)
+    {
+        IHqlExpression * value = localAttr->queryChild(0);
+        if (value)
+        {
+            if (value->isConstant())
+                instance->isLocal = getBoolValue(value, false);
+            else
+                reportError(queryLocation(expr), ECODETEXT(HQLERR_AttributeXMustBeConstant), "LOCAL");
```

Comment:
Should this be done in the semantic check phase?

Reply:
I have moved it, although I'm not convinced it is better.